src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.XPlat/UpdateNuGetConfigPackageSourcesMappings.cs

```diff
@@ -248,19 +255,39 @@ private void DiscoverPackagesFromSBRPCacheSource(XElement pkgSourcesElement, Has
             }
         }
 
-        private void AddToHashtable(Hashtable hashtable, string key, string value)
+        private void GetExistingFilteredSourceMappings(XElement pkgSrcMappingElement, Dictionary<string, List<string>> currentPackages, Dictionary<string, List<string>> referencePackages, Dictionary<string, List<string>> previouslyBuiltPackages, Dictionary<string, List<string>> oldSourceMappingPatterns)
         {
-            if (hashtable.ContainsKey(key))
+            foreach (XElement packageSource in pkgSrcMappingElement.Descendants().Where(e => e.Name == "packageSource"))
             {
-                List<string> values = (List<string>)hashtable[key];
+                List<string> filteredPatterns = new List<string>();
+                foreach (XElement package in packageSource.Descendants().Where(e => e.Name == "package"))
+                {
+                    string pattern = package.Attribute("pattern").Value.ToLower();
+                    if (!currentPackages.ContainsKey(pattern) &&
+                        !referencePackages.ContainsKey(pattern) &&
+                        !previouslyBuiltPackages.ContainsKey(pattern))
+                    {
+                        filteredPatterns.Add(pattern);
+                    }
+                }
+
+                oldSourceMappingPatterns.Add(packageSource.Attribute("key").Value, filteredPatterns);
+            }
+        }
+
+        private void AddToDictionary(Dictionary<string, List<string>> dictionary, string key, string value)
+        {
+            if (dictionary.ContainsKey(key))
```

Comment:
TryGetValue looks to be a better choice here as it avoids indexing the dictionary multiple times.

Reply:
Will fix, for simpler code (one less line). Not sure about perf as `TryGetValue` essentially invokes `ContainsKey` and does the indexing.

Reply:
Fixed with https://github.com/dotnet/installer/pull/19114/commits/3f458e5268b5925b8fae42f35138b3f344b24f28